File: matatu-sim-python/member4_road_rendering.py


import pygame

from member1_setup_config import (
    BASE_W, BASE_H, ROAD_X, ROAD_W, ROAD_RIGHT, LANE_COUNT, LANE_W, C,
    MATATU_W, MATATU_H, MATATU_Y,
    rounded_rect, circle, draw_text,
)
from member2_matatu_controls import state, matatu_rect

# All drawing happens on this fixed-size "base" surface, which is then
# scaled and letterboxed onto the real window (mirrors the JS canvas scale/translate logic).
base_surface = pygame.Surface((BASE_W, BASE_H))


def draw_road(surface):
    pygame.draw.rect(surface, C.roadShadow, (ROAD_X - 10, 0, ROAD_W + 20, BASE_H))
    pygame.draw.rect(surface, C.road, (ROAD_X, 0, ROAD_W, BASE_H))

    pygame.draw.rect(surface, C.white, (ROAD_X + 12, 0, 5, BASE_H))
    pygame.draw.rect(surface, C.white, (ROAD_RIGHT - 16, 0, 5, BASE_H))

    for lane_line in range(1, LANE_COUNT):
        x = ROAD_X + lane_line * LANE_W - 3
        y = -80 + state.road_offset
        while y < BASE_H:
            rounded_rect(surface, x, y, 6, 44, 3, C.yellow)
            y += 80

    rounded_rect(surface, ROAD_X + 30, BASE_H - 34, ROAD_W - 60, 26, 13, (0x20, 0x25, 0x2b))
    draw_text(surface, "AUTONOMOUS LANE CAMERA SIMULATION", ROAD_X + ROAD_W / 2, BASE_H - 27, 15, C.muted, "400", "center")


def draw_obstacle(surface, obstacle):
    if obstacle["kind"] == "VEHICLE":
        x, y, w, h = obstacle["x"], obstacle["y"], obstacle["w"], obstacle["h"]
        rounded_rect(surface, x, y, w, h, 7, obstacle["color"], (0x82, 0x1e, 0x1e), 2)
        rounded_rect(surface, x + 12, y + 9, w - 24, 17, 3, (0x96, 0xdc, 0xeb))
        rounded_rect(surface, x + 15, y + 36, 18, 12, 2, (0xeb, 0xf5, 0xff))
        rounded_rect(surface, x + w - 33, y + 36, 18, 12, 2, (0xeb, 0xf5, 0xff))
        circle(surface, x + 13, y + h - 10, 8, C.black)
        circle(surface, x + w - 13, y + h - 10, 8, C.black)
        if obstacle["detected"]:
            rounded_rect(surface, x - 5, y - 5, w + 10, h + 10, 8, None, C.red, 3)
        return

    x, y, w, h = obstacle["x"], obstacle["y"], obstacle["w"], obstacle["h"]
    rounded_rect(surface, x, y, w, h, 2, (0x2a, 0x2d, 0x31))
    stripe_x = x + 12
    while stripe_x < x + w - 12:
        rounded_rect(surface, stripe_x, y + 9, 28, h - 18, 3, C.white)
        stripe_x += 44
    rounded_rect(surface, x + w / 2 - 82, y + h / 2 - 14, 164, 28, 4, C.white)
    label = "CLEARED" if obstacle["cleared"] else "ZEBRA CROSSING"
    draw_text(surface, label, x + w / 2, y + h / 2 - 8, 15, C.black, "700", "center")

    if obstacle["cleared"]:
        rounded_rect(surface, x - 1, y - 3, w + 2, h + 6, 4, None, C.green, 3)
    elif obstacle["detected"]:
        rounded_rect(surface, x - 1, y - 3, w + 2, h + 6, 4, None, C.red, 3)


def draw_matatu(surface):
    matatu = matatu_rect()
    mx, my = matatu.x, matatu.y
    rounded_rect(surface, mx + 8, my + 10, MATATU_W, MATATU_H, 12, (0, 0, 0))
    rounded_rect(surface, mx, my, MATATU_W, MATATU_H, 12, C.matatuYellow, (0x8c, 0x68, 0x0a), 3)
    rounded_rect(surface, mx + 10, my + 14, MATATU_W - 20, 38, 8, C.matatuBlue)
    rounded_rect(surface, mx + 17, my + 20, MATATU_W - 34, 26, 5, C.cyan)

    for i in range(3):
        rounded_rect(surface, mx + 12 + i * 22, my + 62, 16, 18, 3, (0xa6, 0xe5, 0xee))

    rounded_rect(surface, mx + 16, my + MATATU_H - 28, 18, 10, 3, (0xff, 0xee, 0x75))
    rounded_rect(surface, mx + MATATU_W - 34, my + MATATU_H - 28, 18, 10, 3, (0xff, 0xee, 0x75))
    circle(surface, mx + 14, my + 34, 8, C.black)
    circle(surface, mx + MATATU_W - 14, my + 34, 8, C.black)
    circle(surface, mx + 14, my + MATATU_H - 24, 8, C.black)
    circle(surface, mx + MATATU_W - 14, my + MATATU_H - 24, 8, C.black)

    rounded_rect(surface, mx, my - 16, MATATU_W, 8, 4, C.red if state.braking else C.green)
    draw_text(surface, "ISUZU PSV", mx + MATATU_W / 2, my + 94, 15, C.black, "700", "center")


def draw_simulation(screen):
    from member5_dashboard_camera import draw_left_panel, draw_right_panel, draw_game_over

    base_surface.fill(C.bg)

    draw_road(base_surface)
    for obstacle in state.obstacles:
        draw_obstacle(base_surface, obstacle)
    draw_matatu(base_surface)
    draw_left_panel(base_surface)
    draw_right_panel(base_surface)
    if state.game_over:
        draw_game_over(base_surface)

    win_w, win_h = screen.get_size()
    scale = min(win_w / BASE_W, win_h / BASE_H)
    scaled_w, scaled_h = int(BASE_W * scale), int(BASE_H * scale)
    ox, oy = (win_w - scaled_w) // 2, (win_h - scaled_h) // 2

    screen.fill(C.bg)
    scaled = pygame.transform.smoothscale(base_surface, (scaled_w, scaled_h))
    screen.blit(scaled, (ox, oy))

File: member4_road_rendering.cpp

#include "member4_road_rendering.h"
#include "member5_dashboard_camera.h"
#include <algorithm>

void drawRoad(sf::RenderTarget& target) {
    sf::RectangleShape shadow({ ROAD_W + 20.f, BASE_H });
    shadow.setPosition(ROAD_X - 10.f, 0.f);
    shadow.setFillColor(C::roadShadow);
    target.draw(shadow);

    sf::RectangleShape road({ ROAD_W, BASE_H });
    road.setPosition(ROAD_X, 0.f);
    road.setFillColor(C::road);
    target.draw(road);

    sf::RectangleShape leftEdge({ 5.f, BASE_H });
    leftEdge.setPosition(ROAD_X + 12.f, 0.f);
    leftEdge.setFillColor(C::white);
    target.draw(leftEdge);

    sf::RectangleShape rightEdge({ 5.f, BASE_H });
    rightEdge.setPosition(ROAD_RIGHT - 16.f, 0.f);
    rightEdge.setFillColor(C::white);
    target.draw(rightEdge);

    for (int laneLine = 1; laneLine < LANE_COUNT; ++laneLine) {
        float x = ROAD_X + laneLine * LANE_W - 3.f;
        for (float y = -80.f + state.roadOffset; y < BASE_H; y += 80.f) {
            roundedRectFill(target, x, y, 6.f, 44.f, 3.f, C::yellow);
        }
    }

    roundedRectFill(target, ROAD_X + 30.f, BASE_H - 34.f, ROAD_W - 60.f, 26.f, 13.f, C::panelDark);
    drawText(target, "AUTONOMOUS LANE CAMERA SIMULATION", ROAD_X + ROAD_W / 2.f, BASE_H - 27.f,
              15, C::muted, false, Align::Center);
}

void drawObstacle(sf::RenderTarget& target, const Obstacle& obstacle) {
    float x = obstacle.x, y = obstacle.y, w = obstacle.w, h = obstacle.h;

    if (obstacle.kind == "VEHICLE") {
        roundedRect(target, x, y, w, h, 7.f, &obstacle.color, &C::vehicleTrim, 2.f);
        roundedRectFill(target, x + 12.f, y + 9.f, w - 24.f, 17.f, 3.f, C::glassBlue);
        roundedRectFill(target, x + 15.f, y + 36.f, 18.f, 12.f, 2.f, C::headlight);
        roundedRectFill(target, x + w - 33.f, y + 36.f, 18.f, 12.f, 2.f, C::headlight);
        drawCircle(target, x + 13.f, y + h - 10.f, 8.f, C::black);
        drawCircle(target, x + w - 13.f, y + h - 10.f, 8.f, C::black);
        if (obstacle.detected) {
            roundedRectOutline(target, x - 5.f, y - 5.f, w + 10.f, h + 10.f, 8.f, C::red, 3.f);
        }
        return;
    }

    roundedRectFill(target, x, y, w, h, 2.f, C::asphaltDark);
    for (float stripeX = x + 12.f; stripeX < x + w - 12.f; stripeX += 44.f) {
        roundedRectFill(target, stripeX, y + 9.f, 28.f, h - 18.f, 3.f, C::white);
    }
    roundedRectFill(target, x + w / 2.f - 82.f, y + h / 2.f - 14.f, 164.f, 28.f, 4.f, C::white);
    drawText(target, obstacle.cleared ? "CLEARED" : "ZEBRA CROSSING", x + w / 2.f, y + h / 2.f - 8.f,
              15, C::black, true, Align::Center);

    if (obstacle.cleared) {
        roundedRectOutline(target, x - 1.f, y - 3.f, w + 2.f, h + 6.f, 4.f, C::green, 3.f);
    } else if (obstacle.detected) {
        roundedRectOutline(target, x - 1.f, y - 3.f, w + 2.f, h + 6.f, 4.f, C::red, 3.f);
    }
}

void drawMatatu(sf::RenderTarget& target) {
    sf::FloatRect matatu = matatuRect();
    float mx = matatu.left, my = matatu.top;

    roundedRectFill(target, mx + 8.f, my + 10.f, MATATU_W, MATATU_H, 12.f, sf::Color(0, 0, 0, 153));
    roundedRect(target, mx, my, MATATU_W, MATATU_H, 12.f, &C::matatuYellow, &C::matatuTrim, 3.f);
    roundedRectFill(target, mx + 10.f, my + 14.f, MATATU_W - 20.f, 38.f, 8.f, C::matatuBlue);
    roundedRectFill(target, mx + 17.f, my + 20.f, MATATU_W - 34.f, 26.f, 5.f, C::cyan);

    for (int i = 0; i < 3; ++i) {
        roundedRectFill(target, mx + 12.f + i * 22.f, my + 62.f, 16.f, 18.f, 3.f, C::windowTint);
    }

    roundedRectFill(target, mx + 16.f, my + MATATU_H - 28.f, 18.f, 10.f, 3.f, C::lampYellow);
    roundedRectFill(target, mx + MATATU_W - 34.f, my + MATATU_H - 28.f, 18.f, 10.f, 3.f, C::lampYellow);
    drawCircle(target, mx + 14.f, my + 34.f, 8.f, C::black);
    drawCircle(target, mx + MATATU_W - 14.f, my + 34.f, 8.f, C::black);
    drawCircle(target, mx + 14.f, my + MATATU_H - 24.f, 8.f, C::black);
    drawCircle(target, mx + MATATU_W - 14.f, my + MATATU_H - 24.f, 8.f, C::black);

    roundedRectFill(target, mx, my - 16.f, MATATU_W, 8.f, 4.f, state.braking ? C::red : C::green);
    drawText(target, "ISUZU PSV", mx + MATATU_W / 2.f, my + 94.f, 15, C::black, true, Align::Center);
}

void drawSimulation() {
    sf::RenderTexture& canvas = baseCanvas();
    canvas.clear(C::bg);

    drawRoad(canvas);
    for (const auto& obstacle : state.obstacles) drawObstacle(canvas, obstacle);
    drawMatatu(canvas);
    drawLeftPanel(canvas);
    drawRightPanel(canvas);
    if (state.gameOver) drawGameOver(canvas);

    canvas.display();

    sf::RenderWindow& window = gameWindow();
    sf::Vector2u winSize = window.getSize();
    float scale = std::min(winSize.x / BASE_W, winSize.y / BASE_H);
    float scaledW = BASE_W * scale, scaledH = BASE_H * scale;
    float ox = (winSize.x - scaledW) / 2.f, oy = (winSize.y - scaledH) / 2.f;

    window.clear(C::bg);

    sf::Sprite sprite(canvas.getTexture());
    sprite.setScale(scale, scale);
    sprite.setPosition(ox, oy);
    window.draw(sprite);
}
